### PR TITLE
feat: Implement basic version of string to float/double/decimal

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Related to https://github.com/apache/datafusion-comet/issues/326 and https://github.com/apache/datafusion-comet/issues/325

## Rationale for this change

When running aggregate queries against schemas where numbers are in string format (csv files, parquet files generated from csv files, etc), Spark will add casts from string to floating point or decimal. We fall back to Spark because we did not have these casts implemented.

This is frustrating for casual users who just want to try Comet out on some text inputs.

## What changes are included in this PR?

This PR adds a basic implementation of these casts by delegating to DataFusion. These are not compatible with Spark and are documented as such and are only enabled if `spark.comet.cast.allowIncompatible` is enabled.

I recommend viewing the diff without whitespace changes (https://github.com/apache/datafusion-comet/pull/870/files?diff=split&w=1)

## How are these changes tested?

New tests

### CODE REVIEW

File: native/spark-expr/src/cast.rs [hunk 1]
```diff
@@ -142,6 +142,8 @@ pub struct Cast {
     /// When cast from/to timezone related types, we need timezone, which will be resolved with
     /// session local timezone by an analyzer in Spark.
     pub timezone: String,
+
+    pub allow_incompat: bool,
```
Comment (comphead) [inline native/spark-expr/src/cast.rs:146]: lets have a comment on this field

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 2]
```diff
@@ -555,17 +555,51 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     castTest(gen.generateStrings(dataSize, numericPattern, 8).toDF("a"), DataTypes.FloatType)
   }
 
+  test("cast StringType to FloatType (partial support)") {
```
Comment (comphead) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:558]: may be we can have another test showing what is not supported?

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:558]: we do have failing tests already that are ignored:

```scala
  ignore("cast StringType to FloatType") {
    // https://github.com/apache/datafusion-comet/issues/326
    castTest(gen.generateStrings(dataSize, numericPattern, 8).toDF("a"), DataTypes.FloatType)
  }
  
    ignore("cast StringType to DoubleType") {
    // https://github.com/apache/datafusion-comet/issues/326
    castTest(gen.generateStrings(dataSize, numericPattern, 8).toDF("a"), DataTypes.DoubleType)
  }

  ignore("cast StringType to DecimalType(10,2)") {
    // https://github.com/apache/datafusion-comet/issues/325
    val values = gen.generateStrings(dataSize, numericPattern, 8).toDF("a")
    castTest(values, DataTypes.createDecimalType(10, 2))
  }
```

File: native/spark-expr/src/cast.rs (hunk 1 above)
Comment (andygrove) [reply native/spark-expr/src/cast.rs:147]: added

**Review by comphead — APPROVED** [review]
I feel it should be good, considering DF already supports such casts. However I'm thinking if we should have a new variable or use Spark Comet property? Spark Comet property probably is easier to drop when we dont need this logic anymore

**Review by viirya — APPROVED** [review]

**Comment by andygrove** [comment]
> I feel it should be good, considering DF already supports such casts. However I'm thinking if we should have a new struct field `allow_noncompat` or introduce new Spark Comet property? Spark Comet property probably is easier to drop when we dont need this logic anymore

@comphead  The new stuct field is populated via the config `spark.comet.cast.allowIncompatible`. Are you saying that we should pass this config explicitly in JNI rather than add the `allow_noncompat` flag to the protobuf and struct?

File: native/spark-expr/src/cast.rs [hunk 3]
```diff
@@ -711,7 +728,11 @@ fn cast_array(
 
 /// Determines if DataFusion supports the given cast in a way that is
 /// compatible with Spark
```
Comment (huaxingao) [inline native/spark-expr/src/cast.rs:730]: nit: Shall we update the documentation comment to reflect the addition of the `allow_incompat`parameter?

**Review by huaxingao — APPROVED** [review]
LGTM. Thanks for the PR @andygrove

File: native/spark-expr/src/cast.rs (hunk 3 above)
Comment (andygrove) [reply native/spark-expr/src/cast.rs:730]: `allow_incompat` is an internal API so I don't think we need to add antyhing to the user guide. We do already have documentation for the `spark.comet.cast.allowIncompatible` config, which is used to populate `allow_incompat`.

**Comment by andygrove** [comment]
Thanks for the reviews @comphead @viirya @huaxingao

**Pull request merged**